File: backend/src/main/java/com/liaw/dev/cinemaflix/entity/Movie.java
package com.liaw.dev.cinemaflix.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Data
@Entity
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "tb_movie")
public class Movie {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String title;
    private String description;
    @Column(name = "synopsis", length = 4000)
    private String synopsis;
    private Integer duration;
    private Integer rating;
    private Integer age_rating;
    private LocalDate publication_year;
    private Boolean top10 = false;

    @ManyToMany
    @JoinTable(
            name = "movie_category",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private List<Category> categories;

    @ManyToMany
    @JoinTable(
            name = "movie_streaming",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private List<Streaming> streamings;

    @ManyToMany
    @JoinTable(
            name = "movie_actor",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "actor_id")
    )
    private List<Actor> actors;

    @ManyToMany
    @JoinTable(
            name = "movie_language",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "language_id")
    )
    private List<Language> languages;

    @ManyToOne
    @JoinColumn(name = "user_id")
    private User user;

}

File: frontend/src/app/interfaces/movie.ts
import { UserRequest } from "../service/auth.service";
import { ProfileInter } from "./profile";

export interface Movies{
    id:number;
    title:string;
    description:string;
    duration:number;
    rating:number;
    age_rating:number;
    publication_year: number;
    categoryIds:number[];
    streamingIds:number[];
    top10: boolean; 
    isWatchlisted: boolean;
    user:UserRequest;
}

export interface CreateMovies {
    title: string;
    description: string;
    synopsis: string;
    duration: number;
    rating: number;
    age_rating: number;
    publication_year: number;
    categories: Cateogry[];   
    streamings: Streaming[];
    actors: Actor[];          
    languages: Language[];     
    top10: boolean;
}


export interface MoviesIndividual{
    id:number;
    title:string;
    description:string;
    synopsis:string;
    duration:number;
    rating:number;
    age_rating:number;
    publication_year: number;
    categories: Cateogry[];
    streamings: Streaming[]; 
    actors: Actor[];      
    languages: Language[];
    top10:boolean;
    user:ProfileInter;
}

export interface Page<T> {
  content: T[];           
  totalElements: number;  
  totalPages: number;      
  number: number;           
  size: number;            
  first: boolean;         
  last: boolean;           
}

export interface Cateogry{
    id:number;
    name:string;
}

export interface Actor{
    id:number;
    name:string;
}

export interface Language{
    id:number;
    name:string;
}

export interface Streaming{
    id:number;
    name:string
}

